Budget and specific discharge when stresses applied along boundaries

etiennebresciani opened the pull request:
One may want to allow flow through boundaries instead of treating boundary stresses as an internal source/sink (behavior until now). This is made possible by new post-processing and plotting functions.

The idea is best illustrated with a simple example from tutorial 1. Until now, plotting the specific discharge would give:
<img src="https://raw.githubusercontent.com/etiennebresciani/share/master/tutorial1_old.png" width=400 height=400>
Notice that the arrows in the first and last columns are half of the other ones; this is because boundary faces are implicitly always treated as no flow. With what I developed, we can assign the flow from/to constant head cells to boundary faces, and thus get:
<img src="https://raw.githubusercontent.com/etiennebresciani/share/master/tutorial1_new.png" width=400 height=400>
The essential lines to get the latter figure with the new functions are:
```python
# generate boundary faces information
consthead_ifaces = []
xlow_iface = 1
xhigh_iface = 2
for row in range(nrow):
    consthead_ifaces.append([0, row, 0, xlow_iface])
    consthead_ifaces.append([0, row, ncol-1, xhigh_iface])
boundary_ifaces = {'CONSTANT HEAD': consthead_ifaces}

# get specific discharge
qx, qy, qz = flopy.utils.postprocessing.get_specific_discharge(mf, 
                                    cbcfile, boundary_ifaces=boundary_ifaces, 
                                    hdsfile=hdsfile)

# plot
modelmap.plot_vector(qx, qy, color='white')
```
I tried my best to account for all possible stress packages and MODFLOW versions. However, the new functionality is not available for MODFLOW 6. It would be quite complicated since the cell faces can be arbitrary. Thus, for MODFLOW 6, I simply take the DATA-SPDIS vector from the budget file.

PS1: In line with the proposed developments, I am suggesting the two-steps workflow 1) get_specific_discharge() and 2) plot_vector() as the new workflow for plotting specific discharge in all cases. This is better in line with the existing workflow for heads (i.e., 1) get_data() and 2) plot_array()), and so I think this is an improvement.

PS2: As a consequence, I deprecated a few functions which are now redundant since the new functions are just more flexible. In fact, the previous functionalities are now coded more concisely and generically. In particular, `plot_specific_discharge()` (for “classical” MODFLOW versions) and `plot_discharge()` (for MODFLOW 6) are replaced by a generic `plot_vector()` function (in both `PlotMapView` and `PlotCrossSection` classes).

PS3: A few commits in this PR relate to various fixes that I made along the way but are in fact independent of the above-stated intention. I tried my best for writing explicit commit messages.

PS4: Hope you will find it useful! Please let me know if you have any question or concern.

langevin-usgs commented:
Hi @etiennebresciani, thanks for putting this together.  Looks like there are some style issues and a weird tricontour error for Python 3.5.  Can you try to make revisions and get all the checks to pass?  The concept for improving the discharge vectors makes sense, especially for small models where the vectors along edges clearly have problems.

langevin-usgs commented on autotest/t007_test.py:
```diff
@@ -172,8 +172,8 @@ def test_output_helper_shapefile_export():
     flopy.export.utils.output_helper(os.path.join('temp', 'test.shp'), ml,
-                                     {'HDS': head, 'cbc': cbc},
-                                     mflay=1, kper=10)
+                                      {'HDS': head, 'cbc': cbc},
```
There are changes scattered throughout this PR where a continued line is indented further by one space.  I think we want to align with opening delimiter (https://www.python.org/dev/peps/pep-0008/#indentation), so I don't think we want these indentation changes.

etiennebresciani replied:
This was an unwanted result of commenting and uncommenting bocks in Spyder editor; do not know why Spyder has this weird indentation habit... anyways, I corrected these.

langevin-usgs commented on examples/data/postproc_test/t070_mf2005/t070_mf2005.bas:
```diff
@@ -0,0 +1,43 @@
+# BAS6 package for  MODFLOW-2005, generated by Flopy.
```
You've included input and output files to be versioned within the flopy repo.  We try to avoid this if at all possible.  If files are to be versioned, then they should only be versioned if necessary.  For example, there should be no need to version this list files, the binary grid file, etc.  Or is there?

etiennebresciani replied:
I need most of them for the new tests I set up, but I agree that not all of them are needed (in particular, the .lst and .grb are indeed not needed).

If you think it is better, I could also just generate the model input and output files as part of the tests, and thus simply use the autotest/temp directory for that. I thought I would rather not do this in order to a have more focused testing, but I am happy to do this if you think it is better. If so, I could remove all the input and output files that I added to the repository.

langevin-usgs replied:
@etiennebresciani, I do think it would be better to generate these input and output files instead of storing them in the flopy repository.  Would you mind making that change?

jlarsen-usgs commented:
@etiennebresciani 

The `plot_vector` code and the `plot_specific_discharge` code are quite similar, except that the `plot_vector` accepts qx, qy, qz, where the `plot_specific_discharge` code accepts a prepared recarray of  ['qx`, `qy`, `qz`] for each model node or a list of recarrays. Because we can use `plot_specific_discharge` directly with modflow 6 specific discharge output that is written to the CBC, it makes more sense to me to keep that around instead of deprecating it. Instead of adding another method to the plotting library, the same result could be accomplished by updating the postprocessing.get_specific_discharge() method output a numpy recarray with the qx, qy, qz data and then pass it into `plot_specific_discharge`. Is there any reason/use case where this wouldn't make sense?

jlarsen-usgs commented:
@etiennebresciani 
Python 3.5 on travis is failing because in  `PlotMapView.contour_array()` lines 278-287 have been removed:

```
 # workaround for tri-contour nan issue
        # use -2**31 to allow for 32 bit int arrays
        plotarray[np.isnan(plotarray)] = -2**31
        if masked_values is None:
            masked_values = [-2**31]
        else:
            masked_values = list(masked_values)
            if -2**31 not in masked_values:
                masked_values.append(-2**31)
```
and lines 289-294 have been added:
```
        # add NaN values to mask
        if ismasked is None:
            ismasked = np.isnan(plotarray)
        else:
            ismasked += np.isnan(plotarray)
```

Some versions of matplotlib tricontour will not contour arrays (or masked arrays) that contain NaN values. This change should be reverted.

etiennebresciani commented:
> Looks like there are some style issues and a weird tricontour error for Python 3.5. Can you try to make revisions and get all the checks to pass?

I corrected the style issues and the tricontour error for Python 3.5. Thank you @jlarsen-usgs for pointing out the tricontour nan issue. I did not anticipate this. I reverted the changes as you suggested.

etiennebresciani commented:
@jlarsen-usgs
> The `plot_vector` code and the `plot_specific_discharge` code are quite similar, except that the `plot_vector` accepts qx, qy, qz, where the `plot_specific_discharge` code accepts a prepared recarray of ['qx', 'qy', 'qz'] for each model node or a list of recarrays. Because we can use `plot_specific_discharge` directly with modflow 6 specific discharge output that is written to the CBC, it makes more sense to me to keep that around instead of deprecating it. Instead of adding another method to the plotting library, the same result could be accomplished by updating the `postprocessing.get_specific_discharge()` method output a numpy recarray with the qx, qy, qz data and then pass it into `plot_specific_discharge`. Is there any reason/use case where this wouldn't make sense?

I chose to have `get_specific_discharge` return a tuple and `plot_vector` accept a tuple, but we can as well have `get_specific_discharge` return a recarray and `plot_vector` accept a recarray, if you think it is better - I am totally fine with this. The tuple version is simply more in line with what is initially retrieved for classical MODFLOW versions (i.e., frf, fff, flf), but I am happy to make it instead more in line with what is initially retrieved for MODFLOW 6 (i.e., a recarray). (Note: I would still keep the steps of reshaping qx, qy and qz and setting inactive and dry cells to nan in `get_specific_discharge`).

My main point was to merge `plot_discharge` (for classical MODFLOW versions) and `plot_specific_discharge` (for MODFLOW 6) into a single, simpler, while at the same time more generic function (the new `plot_vector` function). The idea is that whatever the MODFLOW version, you will always do the same, namely:
```python
qx, qy, qz = flopy.utils.postprocessing.get_specific_discharge(model, cbcfile)
quiver = modelmap.plot_vector(qx, qy)
```

or, if you will:
```python
qdis = flopy.utils.postprocessing.get_specific_discharge(model, cbcfile)
quiver = modelmap.plot_vector(qdis)
```
where qdis is a recarray.

I hope this clarifies my intention. Please let me know your opinion.

jlarsen-usgs commented:
@etiennebresciani 

Okay. I understand the advantage of using `plot_vector()` vs. using `plot_specific_discharge()`. The new method sounds great. Last question, Have you built in masking for vectors by ibound array, or will the user need to do that prior to sending qx, qy, and/or qz into the `plot_vector` method?

etiennebresciani commented:
@jlarsen-usgs
> Have you built in masking for vectors by ibound array, or will the user need to do that prior to sending qx, qy, and/or qz into the `plot_vector` method?

Not exactly, but I have achieved this by automatically setting inactive/no-flow and also dry cells to NaN in the `get_specific_discharge` method (NaN values are not plotted in `plot_vector`). This gives the maximum flexibility to the `plot_vector` method. Furthermore, I think it is always useful to set inactive/no-flow and dry cells to NaN for subsequent postprocessing, and so my idea was that doing this in the `get_specific_discharge` method will save even more work to the user.

PS: A related question is popping to my mind: I did not add a “masked_values” option to the `plot_vector` method as I do not see the need, but let me know if you think otherwise.

langevin-usgs commented:
@etiennebresciani, if you can make this one last change to eliminate the MODFLOW 6 input and output files that were added to the repo for the testing (and instead generate them on the fly), then I think this is ready to go.  I looked through the PR again today and like what you've done.  Your code is very clean and easy to follow.  I also see that you even caught the specific discharge case for MODFLOW 6 where idomain is used and the size of spdis is not equal to nodes.  Thanks for the contribution!

jlarsen-usgs commented:
@etiennebresciani 

I think including a masked_values argument or kwarg would be a good addition to the method, because it will increase user flexibility. Thanks for your contribution!

etiennebresciani commented:
@langevin-usgs 
> if you can make this one last change to eliminate the MODFLOW 6 input and output files that were added to the repo for the testing (and instead generate them on the fly).

Done.

> I looked through the PR again today and like what you've done. [...]

Glad you like it; thanks for the feedback! Appreciated :-)

PS: Travis tests failed for test t037 only with Python 3.8, but I am guessing this is unrelated to my work.

etiennebresciani commented:
@jlarsen-usgs 
> I think including a masked_values argument or kwarg would be a good addition to the method

Ok, I included a masked_values argument. Actually there was already a kwargs, which is passed to quiver.

> Thanks for your contribution!

My pleasure!
